lib/terraform/runner.rb

```diff
@@ -0,0 +1,293 @@
+          }
+        )
+        # _log.debug("Payload:>\n, #{payload}")
+        http_response = terraform_runner_client['api/stack/create'].post(
+          payload, :content_type => 'application/json'
+        )
+        _log.debug("==== http_response.body: \n #{http_response.body}")
+        _log.info("stack_job for template: #{template_path} running ...")
+        Terraform::Runner::Response.parsed_response(http_response)
+      ensure
+        # cleanup temp zip file
+        FileUtils.rm_rf(zip_file_path) if zip_file_path
+        _log.debug("Deleted #{zip_file_path}")
+      end
+
+      # Retrieve TerraformRunner Stack Job details
+      def retrieve_stack_job(stack_id)
+        payload = JSON.generate(
+          {
+            :stack_id => stack_id
+          }
+        )
+        http_response = terraform_runner_client['api/stack/retrieve'].post(
+          payload, :content_type => 'application/json'
+        )
+        _log.info("==== Retrieve Stack Response: \n #{http_response.body}")
+        Terraform::Runner::Response.parsed_response(http_response)
+      end
+
+      # Cancel/Stop running TerraformRunner Stack Job
+      def cancel_stack_job(stack_id)
+        payload = JSON.generate(
+          {
+            :stack_id => stack_id
+          }
+        )
+        http_response = terraform_runner_client['api/stack/cancel'].post(
+          payload, :content_type => 'application/json'
+        )
+        _log.info("==== Cancel Stack Response: \n #{http_response.body}")
+        Terraform::Runner::Response.parsed_response(http_response)
+      end
+
+      # Wait for TerraformRunner Stack Job to complete
+      def wait_until_completes(stack_id)
+        interval_in_secs = stack_job_interval_in_secs
+        max_time_in_secs = stack_job_max_time_in_secs
+
+        response = nil
+        Timeout.timeout(max_time_in_secs) do
+          _log.debug("Starting wait for terraform-runner/stack/#{stack_id} completes ...")
+          i = 0
+          loop do
+            _log.debug("loop #{i}")
+            i += 1
+
+            response = retrieve_stack_job(stack_id)
+
+            _log.info("status: #{response.status}")
+
+            case response.status
+            when "SUCCESS"
+              _log.debug("Successful! (stack_job/:#{stack_id})")
+              break
+
+            when "FAILED"
+              _log.info("Failed! (stack_job/:#{stack_id} fails!)")
+              _log.info(response.error_message)
+              break
+
+            when nil
+              _log.info("No status! stack_job/:#{stack_id} must have failed, check response ...")
+              _log.info(response.message)
+              break
+            end
+            _log.info("============\n stack_job/:#{stack_id} status=#{response.status} \n============")
+
+            # sleep interval
+            _log.debug("Sleep for #{interval_in_secs} secs")
+            sleep interval_in_secs
+
+            break unless i < 20
```

Comment:
My brain reads this easier as `break if i >= 20`

Reply:
why are we both checking the iterations of the the loop and the max time?  do we need this line?